### ISSUE TITLE
Seed Ratio Problem

### ISSUE THREAD

**Issue by tomaspy** [tatsuhiro-t/aria2#517]
If I set in torrent seed mode the torrent rating to 1.0 in the other side the downloader receive only 99.9% and the aria is disconnect the connection from tracker.

**Comment by tatsuhiro-t** [comment]
This could be rounding error of floating point number.
But in practice, this kind of error is not significant in this seed ratio, since aria2 is one of a client in the swarm.  Probably, you have a specific setup to rely on the exact seed ratio, or some sort?

**Comment by tomaspy** [comment]
I would like to send the file only once that is why I would like to use this setting. I try the download with utorrent, deluge, qbittorrent, transmission client the result is same every time... 

I use this command:

aria2c --check-integrity="true" --bt-enable-lpd=true --no-conf --enable-dht="true" /media/test/Torrent/TORRENT_test.torrent --seed-ratio=1.0 -d /media/test/Torrent

Ill try to increase the seed  ratio number to 1.01 this is works the seed 100% but then not close the connection from tracker. Any Idea?

**Comment by tatsuhiro-t** [comment]
Rounding error is inevitable when we use floating point seed ratio, such as 10.0.
If total length of download gets large, the round error could be worse.
I'm interested in the total length of the torrent which causes this issue.  Can you share it with us?
Seed ratio is not the feature to control the exact the amount of the upload transfer.  It  is there to encourage the seeder to upload at least what it downloaded.  There is some cases client may request the same piece (or its partial block) twice, such as network is down or client crashed or something.

**Comment by tomaspy** [comment]
I wait this question :) I mean the file size... In my test I try to send 4 different size file. 150Mb, 1Gb, 5Gb, 15Gb, and if I set the ratio to 1.0 the client received - in every diff size file - only 99,9% if I set the ratio to 1.01 or 1.0001 or 1.0000000000001 (does not matter) the client received 100% :)

**Comment by tatsuhiro-t** [comment]
Thank you.  I have to ask you which version you are using.  I remember that old version has a bug of this kind of problem.  Did you try the latest version of aria2?

**Comment by tomaspy** [comment]
My program version is 1.19.0 under Ubuntu 15.10

**Comment by tatsuhiro-t** [comment]
Thank you.  Could you apply the following patch, and see the issue still happens?
```diff
diff --git a/src/BtPieceMessage.cc b/src/BtPieceMessage.cc
index 80026b2..91fe1c5 100644
--- a/src/BtPieceMessage.cc
+++ b/src/BtPieceMessage.cc
@@ -175,8 +175,9 @@ size_t BtPieceMessage::getMessageHeaderLength()
 
 namespace {
 struct PieceSendUpdate : public ProgressUpdate {
-  PieceSendUpdate(std::shared_ptr<Peer> peer, size_t headerLength)
-      : peer(std::move(peer)), headerLength(headerLength)
+  PieceSendUpdate(DownloadContext* dctx, std::shared_ptr<Peer> peer,
+                  size_t headerLength)
+      : dctx(dctx), peer(std::move(peer)), headerLength(headerLength)
   {
   }
   virtual void update(size_t length, bool complete) CXX11_OVERRIDE
@@ -187,7 +188,9 @@ struct PieceSendUpdate : public ProgressUpdate {
       length -= m;
     }
     peer->updateUploadLength(length);
+    dctx->updateUploadLength(length);
   }
+  DownloadContext* dctx;
   std::shared_ptr<Peer> peer;
   size_t headerLength;
 };
@@ -218,10 +221,8 @@ void BtPieceMessage::pushPieceData(int64_t offset, int32_t length) const
   if (r == length) {
     getPeerConnection()->pushBytes(
         buf.release(), length + MESSAGE_HEADER_LENGTH,
-        make_unique<PieceSendUpdate>(getPeer(), MESSAGE_HEADER_LENGTH));
-    // To avoid upload rate overflow, we update the length here at
-    // once.
-    downloadContext_->updateUploadLength(length);
+        make_unique<PieceSendUpdate>(downloadContext_, getPeer(),
+                                     MESSAGE_HEADER_LENGTH));
   }
   else {
     throw DL_ABORT_EX(EX_DATA_READ);
```

**Comment by tomaspy** [comment]
Hi

Happy New Year!

I update the program with the patch. This pacth is working with the large file, but with the small file does not. I try to send 100, 200, 500, 700 Mb files and the downloader receive only 99.9%.
For big files upper 1Gb receive 100%.

**Comment by tatsuhiro-t** [comment]
If peer downloads pieces from multiple peers, then it may request same piece from different peers, to accelerate the last stage of the download (end game mode).  For duplicated pieces, peer will send cancel message.  If that happened, seeder may overcommit upload size, but it is completely normal.  So, I think there is no way to achieve your goal in this case.

**Comment by tomaspy** [comment]
OK... I understand your explanation but not. If I use other software like qbittorrent, utorrent, deluge this problem does not happend I can send small size file with 1.0 ratio, but for me the gui is not good only the CLI that is why I choose the aria2...

In this situation only ONE seeder and ONE receiver available not more. And why working if I seed large size file? If I seed 331.3 Mb (331.306.368 byte) why not send the 100%? And why send 3,8Gb  in 100%? This ratio means:  download X MB file size and  upload same X MB and then the ratio 1.0 and/or if I want to share a file only once for one downloader I can control with ratio how many times can download it. If the receiver has download the file 100% the aria is disconnect from tracker and the receiver never can download again for more or other download they need to request for new torrent file.

So is it possible to add a "after x.x ratio wait X second before disconnect from tracker?" Maybe if the aira is connected to the tracker for X second after the 1.0 ratio they can send it the missing part. What do you think?

**Comment by tatsuhiro-t** [comment]
Since I have not reproduced this so far in my environment, I have no idea why this happens.
I tested 10m torrent, and did many tests, but they all succeeded; 100 % download.
I don't want to add said option, since share ratio is really not the feature intended to be used in this way.

**Comment by tomaspy** [comment]
Experience why not work the seed ratio for small file:
If any torrent creator create the file split for a small file (under 1Gb) - in automatic mode - the creator use 128-768kb split. This split is too small for seeding because the seed ratio control does not watch how many package was sent to the client, they watch only the file size, Is the file size is match (ratio 1.0) the aria is disconnect from the tracker and the end is loss 2 or 3 package(split)  that is why sent only 99.92%.

If we incrase the split in manual mode for small size file to 1024kb (or bigger) the seed ratio control is 100% working the seeder sent the all package.

**Comment by tatsuhiro-t** [comment]
Thank you for the news.
The interesting part is that I used 256KiB piece size, and total 315MiB single file torrent, and could not reproduce this issue.
And aria2 actually checks how many bytes are uploaded against completed size (this seed ratio check gets started when torrent download finished, so completed size == total file size.  This may not be true if --select-file is used, and partial downloads are done).
So.. are you suggesting that we should use the number of pieces we uploaded, instead of the number of bytes uploaded?

**Comment by tomaspy** [comment]
From the torrent file we know the package/piece number and one package size in byte, we know all info vice versa, 2500x1024Kib = 2,560,000 Kib=1.0 Ratio (the last package/piece is always smaller if the file size is not equal to the spliter size.)

If we want to download or seed, we send and receive not just bytes, packages too.

1 package/piece upload/download
IF package/piece                                (2500)
upload/download -OK or Finish
THEN                                                  (1024Kib)
upload/download size is XMb
THEN                                                   (2,560,000 Kib)
our seed ratio is X.X                            1.0 = 2500 piece

**Comment by tatsuhiro-t** [comment]
Do you mean package == piece?

I'm not sure I fully understand your comment, but probably you insisted that number of pieces should be used to calculate share ratio.
It is different from the share I know of.  My understanding is that it just the number of bytes uploaded.

From https://en.wikipedia.org/wiki/Glossary_of_BitTorrent_terms#Share_ratio

```
Share ratio

    A user's share ratio for any individual torrent is a number determined by dividing the amount of data that user has uploaded by the amount of data they have downloaded. Final share ratios over 1.0 carry a positive connotation in the BitTorrent community, because they indicate that the user has sent more data to other users than they received. Likewise, share ratios under 1 have negative connotation. 
```

**Comment by tomaspy** [comment]
Sorry for late reply.. yes package=piece.
And your right the number of pieces equal the ratio. I read the wiki and this is clear but not clear... but never mind now
The seed radio problem is fixed with the bigger piece size.

I have one more question about the seed. How can I configure the aria2 in seed only mode... I mean I create a xxxx.torrent file and I would like to seed only. If I use this command:
aria2c --check-integrity="false" --bt-enable-lpd=true --seed-ratio=1.0 --no-conf --enable-dht="true /home/test/test.torrent -d /home/test/
With this command the aria2 is not seed just listen,
I mean not seeder just peer. If they get any request then seed like a peer.
And I did not find any switch to seed only mode like in GUI torrent clients (after the torrent file creation seed file immediately) there I see the torrent in seed mode.
Any idea?

**Comment by tatsuhiro-t** [comment]
Use -V or --bt-seed-unverified option.  Also make sure that directory specified in -d option contains fully seeded files.

**Comment by tomaspy** [comment]
Thank you, it works fine... one more and I hope last question. If I seed a torrent with utorrent, transmission etc etc and then I push the button then the client connect to the tracker and the tracker print out the hash value of the torrent. If I seed with aria2c I did not see the hash value after the tracker connection. required for this to use a switch? I did not find.

**Comment by tatsuhiro-t** [comment]
If I understand correctly, tracker does some logging, and prints some hash value?   Which hash value it prints out?  info-hash?
What kind of tracker are you using?

**Comment by tomaspy** [comment]
Print out the info hash
udpt tracker

**Comment by tatsuhiro-t** [comment]
Thank you.  Did you see that aria2 does not get any response from tracker?  Or is it just hash value missing from console output?

**Comment by tomaspy** [comment]
If I connect to the tracker with utorrent, transmission etc etc I see in the tracker consol
1 client connect

a4075bbd516a7d9736739771e92668d9


if aria connect
1 client connect

no hash

**Comment by tatsuhiro-t** [comment]
And aria2 got good response from tracker, yes?

**Comment by tomaspy** [comment]
UDPT received CONNECT reply from x.x.x.x:6969 transaction_id=1001331254,connection_id=-4763401030874431489

UDPT sent ANNOUNCE to x.x.x.x:6969 transaction_id=811394366, connection_id=-4763401030874431489, event=STARTED, infohash=55183224304b99cda891166cf2b5b4bbab320846

UDPT received ANNOUNCE reply from x.x.x.x:6969 transaction_id=811394366,connection_id=-4763401030874431489, event=STARTED, infohash=55183224304b99cda891166cf2b5b4bbab320846, interval=1800, leechers=0, seeders=0, num_peers=0

**Comment by tatsuhiro-t** [comment]
I checked UDP tracker specification, and we are sure that we sent all required fields to tracker: http://www.bittorrent.org/beps/bep_0015.html

One possibility is scrape request, which aria2 does not perform.  If other clients do, then the output of udp tracker may be caused by this.  Still I'm not sure about that.

Anyway, as long as UDP tracker works, I think this is not an urgent problem.

**Comment by tomaspy** [comment]
OK thanks for your feedback.. but the problem is aria2c is connected with the tracker but if other client is connect with download request (peer)  aria2c is not seed, just wait...
Aria2c is start seeding If I seed on same machine the same torrent with other software... enough only the first 1 bite. From that point aria2c is seed anytime that torrent.
Ill try with udpt tracker and Opentracker too.  Same result.

**Comment by tatsuhiro-t** [comment]
Right, I found a bug which prevents seeding with udp tracker.  Fix will be committed very soon.

**Comment by tatsuhiro-t** [comment]
Fix committed via ddb94eb.  This will make things easier.
